packages/relay-compiler/transforms/RelayMaskTransform.js

```diff
@@ -141,17 +145,39 @@ function isUnmaskedSpread(spread: FragmentSpread): boolean {
   return Boolean(spread.metadata && spread.metadata.mask === false);
 }
 
-function areSameArgumentDefinitions(
-  argDef1: ArgumentDefinition,
-  argDef2: ArgumentDefinition,
-) {
-  return (
-    argDef1.kind === argDef2.kind &&
-    argDef1.name === argDef2.name &&
-    isEquivalentType(argDef1.type, argDef2.type) &&
+function findCompatibleArgDef(
+  currentArgDef: ArgumentDefinition,
+  newArgDef: ArgumentDefinition,
+): ?ArgumentDefinition {
+  let currentType = currentArgDef.type;
+  let newType = newArgDef.type;
+  let compatibleArgDef = currentArgDef;
+
+  if (
+    !(currentArgDef.type instanceof GraphQLNonNull) &&
+    newArgDef.type instanceof GraphQLNonNull
+  ) {
+    // Current argDef is not compatible with new argDef
+    newType = newArgDef.type.ofType;
+    compatibleArgDef = newArgDef;
+  } else if (
+    currentArgDef.type instanceof GraphQLNonNull &&
+    !(newArgDef.type instanceof GraphQLNonNull)
+  ) {
+    currentType = currentArgDef.type.ofType;
+  }
+
+  if (
+    currentArgDef.kind === newArgDef.kind &&
+    currentArgDef.name === newArgDef.name &&
+    isEquivalentType(currentType, newType) &&
     // Only LocalArgumentDefinition defines defaultValue
-    (argDef1: any).defaultValue === (argDef2: any).defaultValue
-  );
+    (currentArgDef: any).defaultValue === (newArgDef: any).defaultValue
+  ) {
+    return compatibleArgDef;
+  }
+
+  return null;
```

Comment:
I think this could be simplified quite a bit, to something like:

```javascript
const {isTypeSubTypeOf} = require('graphql');
...
function findCompatibleArgDef(prevArgDef, nextArgDef, schema): ?ArgumentDefinition {
  if (prevArgDef.kind !== nextArgDef.kind || prevArgDef.name !== nextArgDef.name || (prevArgDef: any).defaultValue !== (nextArgDef: any).defaultValue) {
    return null;
  } else if (isTypeSubTypeOf(schema, nextArgDef.type, prevArgDef.type) {
    return prevArgDef;
  } else if (isTypeSubTypeOf(schema, prevArgDef.type, nextArgDef.type) {
    return nextArgDef;
  } else {
    return null;
  }
}

```